### PR TITLE
Task_08 module implementation

### PR DESCRIPTION
**Pull request by vadym-kharchuk**

### CODE REVIEW

File: 08_irq/poll_button/khmod.c [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+{
+	static bool old_button_state;
+	bool button_state;
+	ktime_t kt;
+
+	button_state = gpio_get_value(BUTTON);
+
+	if (old_button_state != button_state) {
+		gpio_set_value(LED_MMC, 1);
+		kt = ktime_set(0, BLINK_TIME_MS * 1000000);
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: if you want to create ktime from miliseconds - better to use ms_to_ktime

File: 08_irq/poll_button/khmod.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		hrtimer_start(&tim_h, kt, HRTIMER_MODE_REL);
+	} else if (!button_state) {
+		gpio_set_value(LED_SD, 1);
+	} else {
+		gpio_set_value(LED_SD, 0);
+	}
+
+	old_button_state = button_state;
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: this code could be simplified
} else {
    gpio_set_value(LED_SD, button_state ? 0 : 1);
}

File: 08_irq/poll_button/khmod.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+static int led_gpio_init(int gpio)
+{
+	int rc;
+
+	rc = gpio_direction_output(gpio, 0);
+	if (rc)
+		return rc;
+
+	return 0;
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: just return rc;

File: 08_irq/poll_button/khmod.c [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	timer_setup(&tim_l, button_poll, 0);
+	add_timer(&tim_l);
+
+	hrtimer_init(&tim_h, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	tim_h.function = &hr_callback;
+
+	return 0;
+
+err_led:
+	button_gpio_deinit();
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c:130]: you init two leds.
what if first - ok, second - fail ?

File: 08_irq/poll_button/khmod.c [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+err_button:
+	return rc;
+}
+
+static void __exit onboard_io_exit(void)
+{
+	hrtimer_cancel(&tim_h);
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: (optional suggestion)
you can do like this
int ret = hrtimer_cancel(&hr_timer);
pr_info("hr_timer was %s", ret ? "active -> cancelled" : "not active");

File: 08_irq/poll_button/khmod.c [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	del_timer_sync(&tim_l);
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: better to use try_to_del_timer_sync

File: 08_irq/irq_button/khmod.c [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	if (rc) {
+		pr_err("Can't set GPIO%d for output\n", LED_SD);
+		goto err_out;
+	}
+
+	rc = led_gpio_init(LED_MMC);
+	if (rc) {
+		pr_err("Can't set GPIO%d for output\n", LED_MMC);
+		goto err_out;
+	}
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c]: need to release IRQ if error happens here

File: 08_irq/irq_button/khmod.c [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+static int led_gpio_init(int gpio)
+{
+	int rc;
+
+	rc = gpio_direction_output(gpio, 0);
+	if (rc)
+		return rc;
+
+	return 0;
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c]: just return rc;

File: 08_irq/irq_button/khmod.c [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+static irqreturn_t button_fn(int irq, void *ptr)
+{
+	struct irq_data_s *data = (struct irq_data_s *)ptr;
+
+	pr_info("IRQ count - %d\n", data->counter);
+
+	if (simulate_busy) {
+		mod_timer(&tim_l, jiffies + msecs_to_jiffies(BLINK_TIME_MS));
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c]: just delay here
timer is not needed

File: 08_irq/irq_button/khmod.c [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+		gpio_set_value(LED_MMC, 1);
+	}
+
+	if (kthread_should_stop())
+		return 0;
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c]: this is not needed here

File: 08_irq/irq_button/khmod.c [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+struct irq_data_s {
+	bool button_state;
+	unsigned int counter;
+	int button_irq;
+};
+
+static int button_gpio = -1;
+
+struct irq_data_s irq_data;
+struct timer_list tim_l;
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c]: timer is not needed

**Review by Anton-Soroka — CHANGES_REQUESTED** [review]
Please implement changes according to comments.

File: 08_irq/poll_button/khmod.c (hunk 4 above)
Comment (vadym-kharchuk) [reply 08_irq/poll_button/khmod.c:130]: Nothing? In led_gpio_init we just set gpio_direction_output as output.

File: 08_irq/poll_button/khmod.c (hunk 5 above)
Comment (vadym-kharchuk) [reply 08_irq/poll_button/khmod.c]: thanks

File: 08_irq/poll_button/khmod.c (hunk 3 above)
Comment (vadym-kharchuk) [reply 08_irq/poll_button/khmod.c]: got it

File: 08_irq/poll_button/khmod.c (hunk 2 above)
Comment (vadym-kharchuk) [reply 08_irq/poll_button/khmod.c]: Yes, right. Thank you!

File: 08_irq/irq_button/khmod.c (hunk 7 above)
Comment (vadym-kharchuk) [reply 08_irq/irq_button/khmod.c]: fixed

File: 08_irq/poll_button/khmod.c [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+{
+	static bool old_button_state;
+	bool button_state;
+	ktime_t kt;
+
+	button_state = gpio_get_value(BUTTON);
+
+	if (old_button_state != button_state) {
+		gpio_set_value(LED_MMC, 1);
+		kt = ms_to_ktime(BLINK_TIME_MS * 1000000);
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: ms_to_ktime have internal multiply by 1000000

Comment (vadym-kharchuk) [reply 08_irq/poll_button/khmod.c]: Ow, miss it

File: 08_irq/poll_button/khmod.c [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+static void __exit onboard_io_exit(void)
+{
+	int ret;
+
+	ret = hrtimer_cancel(&tim_h);
+	pr_info("hr_timer was %s", ret ? "active -> cancelled" : "not active");
+
+	ret = try_to_del_timer_sync(&tim_l);
+	pr_info("hr_timer was %s", ret ? "not active" : "active -> cancelled");
```
Comment (Anton-Soroka) [inline 08_irq/poll_button/khmod.c]: Fix typo in format string - this is "timer_list timer", not a "hr_timer".

File: 08_irq/irq_button/khmod.c [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+err_input:
+	gpio_free(gpio);
+err_register:
+	return rc;
+}
+
+static void button_gpio_deinit(void)
+{
+	if (button_gpio >= 0) {
+		gpio_free(button_gpio);
```
Comment (Anton-Soroka) [inline 08_irq/irq_button/khmod.c:108]: In order to fix bug that sometimes appears message l4_per_cm:clk:009c:0: failed to disable
need to turn off debounce before freeing gpio by calling gpio_set_debounce(GPIO_BUTTON, 0);

**Review by Anton-Soroka — CHANGES_REQUESTED** [review]
Please fix minor bugs (in comments).

**Review by Anton-Soroka — APPROVED** [review]

**Pull request merged**